File: src/GameBoard.kt
import kotlin.random.Random
import Cell.State

class GameBoard(
    private val sideLength: Int,
    boardMode: String = STYLE_RANDOM
) {
    val board: Array<Array<Cell>>

    init {
        when(boardMode) {
            STYLE_SWITCH -> {
                board = Array(sideLength) { y ->
                    kotlin.Array(sideLength) { x ->
                        val state = createBreederState(x, y, GRAPHIC_OFFSET)
                        Cell(x, y, state)
                    }
                }
            }
            else -> {
                board = Array(sideLength) { y ->
                    kotlin.Array(sideLength) { x ->
                        val r = Random.nextInt() % 8
                        val state = if (r < 2) State.Living else State.Dead
                        Cell(x, y, state)
                    }
                }
            }
        }
    }

    private fun createBreederState(x: Int, y: Int, offset: Int)  =
        when {
            x == 1 + offset && y == 1 + offset -> State.Living
            x == 2 + offset && y == 1 + offset -> State.Living
            x == 3 + offset && y == 1 + offset -> State.Living
            x == 5 + offset && y == 1 + offset -> State.Living
            x == 1 + offset && y == 2 + offset -> State.Living
            x == 1 + offset && y == 1 + offset -> State.Living
            x == 4 + offset && y == 3 + offset -> State.Living
            x == 5 + offset && y == 3 + offset -> State.Living
            x == 2 + offset && y == 4 + offset -> State.Living
            x == 3 + offset && y == 4 + offset -> State.Living
            x == 5 + offset && y == 4 + offset -> State.Living
            x == 1 + offset && y == 5 + offset -> State.Living
            x == 3 + offset && y == 5 + offset -> State.Living
            x == 5 + offset && y == 5 + offset -> State.Living
            else -> State.Dead
        }

    fun simulateGameBoard() {
        board.forEveryElementsInOrder(
            {
                it.prepareSimulate(getAroundCellsOfACell(it))
            },
            {
                it.simulate()
            }
        )
    }

    private fun getAroundCellsOfACell(cell: Cell): List<Cell> {
        return getCellOfRelationalPosition(cell,
            when {
                cell.isOnStart && cell.isOnTop -> listOf(5,7,8)
                cell.isOnStart && cell.isOnBottom -> listOf(2,3,5)
                cell.isOnEnd && cell.isOnTop -> listOf(4,6,7)
                cell.isOnEnd && cell.isOnBottom -> listOf(1,2,4)
                cell.isOnStart -> listOf(2,3,5,7,8)
                cell.isOnEnd -> listOf(1,2,4,6,7)
                cell.isOnTop -> listOf(4,5,6,7,8)
                cell.isOnBottom -> listOf(1,2,3,4,5)
                else -> listOf(1,2,3,4,5,6,7,8)
            }
        )
    }

    /**
     * @param positions
     * |1|2|3|
     * |4|■|5|
     * |6|7|8|
     */
    private fun getCellOfRelationalPosition(cell: Cell, positions: List<Int>): List<Cell> {
        return positions.map {
            when (it) {
                1 -> board[cell.y - 1][cell.x - 1]
                2 -> board[cell.y - 1][cell.x]
                3 -> board[cell.y - 1][cell.x + 1]
                4 -> board[cell.y][cell.x - 1]
                5 -> board[cell.y][cell.x + 1]
                6 -> board[cell.y + 1][cell.x - 1]
                7 -> board[cell.y + 1][cell.x]
                8 -> board[cell.y + 1][cell.x + 1]
                else -> board[cell.y][cell.x]
            }
        }
    }

    private fun <T> Array<Array<T>>.forEveryElementsInOrder(
        function1: (element1: T) -> Unit,
        function2: (element2: T) -> Unit
    ) {
        this.forEveryElements(function1)
        this.forEveryElements(function2)
    }

}

File: Nim/gameboard.nim
import cell as cl
import utils
import sequtils, strutils, math, algorithm

type Gameboard* = ref object
    sideLen*: int
    offset*: int
    boardArray: seq[seq[Cell]]

method getBoardArray*(gameboard: Gameboard): seq[seq[Cell]] {.base.} =
    gameboard.boardArray

method getCellOfRelationalPosition(gameboard: Gameboard, cell: Cell, positions: seq[int]): seq[Cell] {.base.} =
    let 
        board = gameboard.boardArray
        x = cell.getX()
        y = cell.getY()
    
    result = positions.mapIt(
        case it:
            of 1:
                board[y - 1][x - 1]
            of 2:
                board[y - 1][x]
            of 3:
                board[y - 1][x + 1]
            of 4:
                board[y][x - 1]
            of 5:
                board[y][x + 1]
            of 6:
                board[y + 1][x - 1]
            of 7:
                board[y + 1][x]
            of 8:
                board[y + 1][x + 1]
            else:
                raise newException(IndexError, "第二引数はseq[int]で1~8の値を持つものにしてください")
    )

method getAroundCells(gameboard: Gameboard, cell: Cell): seq[Cell] {.base.} =
    var aroundPositions: seq[int]
    let
        isOnTop = cell.isOnTop()
        isOnBottom = cell.isOnBottom(gameboard.sideLen)
        isOnStart = cell.isOnStart()
        isOnENd = cell.isOnEnd(gameboard.sideLen)

    if isOnStart and isOnTop:
        aroundPositions = @[5, 7, 8]
    elif isOnStart and isOnBottom:
        aroundPositions = @[2, 3, 5]
    elif isOnEnd and isOnTop:
        aroundPositions = @[4, 6, 7]
    elif isOnEnd and isOnBottom:
        aroundPositions = @[1, 2, 4]
    elif isOnStart:
        aroundPositions = @[2, 3, 5, 7, 8]
    elif isOnEnd:
        aroundPositions = @[1, 2, 4, 6, 7]
    elif isOnTop:
        aroundPositions = @[4, 5, 6, 7, 8]
    elif isOnBottom:
        aroundPositions = @[1, 2, 3, 4, 5]
    else:
        aroundPositions = @[1, 2, 3, 4, 5, 6, 7, 8]

    gameboard.getCellOfRelationalPosition(cell, aroundPositions)

method toNextGen*(gameboard: Gameboard) {.base.} =
    forEachDubleArray[Cell](gameboard.boardArray, proc(cell: Cell) = cell.prepareCell(gameboard.getAroundCells(cell)))
    forEachDubleArray[Cell](gameboard.boardArray, proc(cell: Cell) = cell.toNextGen())

method createBreeder(gameboard: Gameboard, x: int, y:int, offset: int): CellState {.base.} =
    var state: CellState
    if x == 1 + offset and y == 1 + offset:
        state = ALIVE
    elif x == 2 + offset and y == 1 + offset:
        state = ALIVE
    elif x == 3 + offset and y == 1 + offset:
        state = ALIVE
    elif x == 5 + offset and y == 1 + offset:
        state = ALIVE
    elif x == 1 + offset and y == 2 + offset:
        state = ALIVE
    elif x == 4 + offset and y == 3 + offset:
        state = ALIVE
    elif x == 5 + offset and y == 3 + offset:
        state = ALIVE
    elif x == 2 + offset and y == 4 + offset:
        state = ALIVE
    elif x == 3 + offset and y == 4 + offset:
        state = ALIVE
    elif x == 5 + offset and y == 4 + offset:
        state = ALIVE
    elif x == 1 + offset and y == 5 + offset:
        state = ALIVE
    elif x == 3 + offset and y == 5 + offset:
        state = ALIVE
    elif x == 5 + offset and y == 5 + offset:
        state = ALIVE
    else:
        state = DEAD

    return state
        

method onCreate*(gameboard: Gameboard) {.base.} =
    # インスタンスのnew時に呼ぶこと
    # ライフゲームの初期状態の設定 
    let size = gameboard.sideLen
    if size == 0:
        return
    
    gameboard.boardArray = newSeqWith(size, newSeqWith(size, Cell()))
    forEachIndexed[seq[Cell]](
        gameboard.boardArray,
        proc(row: seq[Cell], y: int) = 
            forEachIndexed[Cell](
                row,
                proc (cell: Cell, x: int) =
                    cell.init(
                        x,
                        y,
                        gameboard.createBreeder(x, y, gameboard.offset)
                    )
            )
    )